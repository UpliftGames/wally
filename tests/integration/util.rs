use std::collections::BTreeMap;
use std::path::Path;

use serde::{Deserialize, Serialize};

#[macro_export]
macro_rules! assert_dir_snapshot {
    ( $path:expr ) => {
        let result = $crate::util::read_path($path).unwrap();
        insta::assert_yaml_snapshot!(result);
    };
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Entry {
    File(String),
    Dir(BTreeMap<String, Entry>),
}

pub fn read_path(path: &Path) -> anyhow::Result<Entry> {
    let meta = fs_err::metadata(path)?;

    if meta.is_dir() {
        let children = fs_err::read_dir(path)?
            .map(|dir_entry| {
                let path = dir_entry?.path();
                let name = path.file_name().unwrap().to_str().unwrap().to_owned();
                let entry = read_path(&path)?;

                Ok((name, entry))
            })
            .collect::<anyhow::Result<BTreeMap<String, Entry>>>()?;

        Ok(Entry::Dir(children))
    } else {
        let contents = normalize_line_ends(fs_err::read_to_string(path)?);
        Ok(Entry::File(contents))
    }
}

fn normalize_line_ends(str: String) -> String {
    let mut new = String::with_capacity(str.len() + 1);
    for line in str.lines() {
        new.push_str(line);
        new.push('\n')
    }
    if !str.ends_with('\n') {
        new.pop();
    }
    new
}
