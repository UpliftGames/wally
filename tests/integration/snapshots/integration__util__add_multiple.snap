---
source: tests/integration/util.rs
expression: "fs::read_to_string(project.path().join(\"wally.toml\")).unwrap()"
---
[package]
name = "biff/server-depends-on-shared"
version = "0.1.0"
license = "MIT"
realm = "server"
registry = "test-registries/primary-registry"

[dependencies]
b = "biff/minimal@0.1.0"
a = "biff/minimal-shared@0.1.0"
minimal = "biff/minimal@0.1.0"
transitive_dependency = "biff/transitive-dependency@=0.1.0"

